Add guide for code coverage reporting (#62)
opensearch-project/opensearch-plugins: TESTING.md

<!-- TOC -->

> - [Testing](#testing)

Reviewer: Can you add links to all the subsections of the tree? 
It would be helpful to navigate.

Refer to UPGRADING.md if you need an example.

Author: Thank you, good catch! I will add all the subsections.

Author: Added all the headers so far.

    - [Code Coverage Collecting](#code-coverage-collecting)
    - [Code Coverage Reporting](#code-coverage-reporting)
<!-- TOC -->

## Testing

### Code Coverage Collecting
Code coverage is a measurement of how many lines of code are executed under the test suites.

#### OpenSearch plugin

##### Gradle project
1. Add [JaCoCo Gradle plugin](https://docs.gradle.org/current/userguide/jacoco_plugin.html) to `build.gradle` in your repository. [example](https://github.com/opensearch-project/sql/blob/opensearch-1.0.0.0-rc1/build.gradle#L48)
2. Set Jacoco to export coverage report in `xml` format. [example](https://github.com/opensearch-project/sql/blob/opensearch-1.0.0.0-rc1/build.gradle#L101-L105)
3. Call gradle task "jacocoTestReport" to generate report (`./gradlew jacocoTestReport`) after running your tests. (Or add a `dependsOn` to make "jacocoTestReport" task depends on your test tasks. [example](https://github.com/opensearch-project/index-management/blob/opensearch-1.0.0.0-beta1/build-tools/coverage.gradle))

##### Maven project
Since OpenSearch uses Gradle as the build tool, we don't recommend using Maven.

> Code coverage for a Maven project can be generated by [JaCoCo Maven plugin](https://www.eclemma.org/jacoco/trunk/doc/maven.html), and we also got an [example](https://github.com/opensearch-project/security/blob/v1.0.0.0-rc1/pom.xml#L636).

Reviewer: Minor nit: Just rearranging the wording to sound simple and active.
```suggestion
Code coverage for a Maven project can be generated by [JaCoCo Maven plugin](https://www.eclemma.org/jacoco/trunk/doc/maven.html), and here is an [example](https://github.com/opensearch-project/security/blob/v1.0.0.0-rc1/pom.xml#L636).
```

Author: 😄 Thanks. No problem!

> #### OpenSearch Dashboards plugin

Reviewer: Just add a TBD and then we could add this section later.

Author: I will write the example of using Jest as the testing framework and generate code coverage report, as the existing plugins are using, so that here will not be empty. 😁

Reviewer: Yay!!

Author: It's very easy to get code coverage report when using Jest testing framework, and I added the guide.

> #### Limitation

Reviewer: Can we also add the limitations of JaCoCo plugin?
As far as I know, it doesn't support PowerMock and Integration tests(can you double check on this, as this was the case for AD).

Author: Thanks for your idea! I will add them 😄.

Author: Added the useful tips as far as I know for using JaCoCo.

### Code Coverage Reporting
OpenSearch plugins are using [Codecov](https://about.codecov.io/) for code coverage reporting and analysis. The dashboard can be seen [here](https://app.codecov.io/gh/opensearch-project/).

All the OpenSearch plugins are required to have the following items:

#### Code coverage report upload through CI workflow
1. Generate code coverage report in `xml` format in a CI workflow of Github Actions
2. Upload the report to Codecov by adding a step of [codecov-action](https://github.com/codecov/codecov-action) in the workflow.

> 3. For private repositories, an upload token is needed, visit the repository page of Codecov to get your token: https://codecov.io/gh/opensearch-project/REPO-NAME-HERE. Please use Github Secrets to store the token and not expose to the public.

Reviewer: Can we add a reference to Github secrets documentation and how it could be setup?

Author: I can add one, but I'm doubt if the plugins are required to use the codecov token. AD plugin is not using codecov token in your sample CI workflow, so I thought if that's fine, I don't need to take time in writing guide of Github Secrets. 😏

Reviewer: Haha I didnt mean to add the whole example, just link github secrets documentation, folks should be able to figure out from that doc on how to add a secret to their repository  it if needed.
https://docs.github.com/en/actions/reference/encrypted-secrets

Author: Thanks for finding the link for me! I added that.

4. Set the workflow to be triggered by `push` and `pull request` on `main` and release branches.
5. To see your code coverage results from Codecov, visiting https://codecov.io/gh/opensearch-project/REPO-NAME-HERE after the CI workflow is triggered.

See [CI Workflows](STANDARDS.md#ci-workflows) for example.

#### A status badge in README file of the repository to show the code coverage.  ([example](https://github.com/opensearch-project/index-management#readme))

> Add the follwing line at the header of the README markdown file:

Author: a typo

##### Gradle project
1. Add [JaCoCo Gradle plugin](https://docs.gradle.org/current/userguide/jacoco_plugin.html) to `build.gradle` in your repository. [example](https://github.com/opensearch-project/sql/blob/opensearch-1.0.0.0-rc1/build.gradle#L48)
2. Set Jacoco to export coverage report in `xml` format. [example](https://github.com/opensearch-project/sql/blob/opensearch-1.0.0.0-rc1/build.gradle#L101-L105)
3. Call gradle task "jacocoTestReport" to generate report (`./gradlew jacocoTestReport`) after running your tests. (Or add a `dependsOn` to make "jacocoTestReport" task depending on your test tasks. [example](https://github.com/opensearch-project/index-management/blob/opensearch-1.0.0.0-beta1/build-tools/coverage.gradle))

##### Maven project
Since OpenSearch uses Gradle as the build tool, we don't recommend using Maven.
Code coverage for a Maven project can be generated by [JaCoCo Maven plugin](https://www.eclemma.org/jacoco/trunk/doc/maven.html), and here is an [example](https://github.com/opensearch-project/security/blob/v1.0.0.0-rc1/pom.xml#L636).

##### Limitation

> 1. JaCoCo Gradle plugin can not collect code coverage properly for integration tests (tests that extend `OpenSearchIntegTestCase` or `OpenSearchRestTestCase`, see [here](https://github.com/opensearch-project/OpenSearch/blob/1.0/TESTING.md#base-classes-for-test-cases) to learn more). However, there is a workaround to collect coverage when the integration test is running on a local single node cluster. For your reference, [this file](https://github.com/opensearch-project/index-management/blob/opensearch-1.0.0.0-beta1/build-tools/coverage.gradle) is a custom Gradle pluign to help achieve, and it can be applied conditionally like [here](https://github.com/opensearch-project/index-management/blob/opensearch-1.0.0.0-beta1/build.gradle#L90-L93).

Reviewer: Super nit: 
`pluign` to `plugin`
`gralde` to `gradle`

Author: Oh, thank you! 😳 You are so careful!

Author: Each kind of mistake spelling has got 2 samples, so terrible. 🤡

3. JaCoCo Gradle plugin can not merge code coverage reports from different sub-projects out-of-box, when the code repository contains a [Gralde multi-project build](https://docs.gradle.org/6.6.1/userguide/multi_project_builds.html). If you need the aggrectiated code coverage report from multiple sub-projects, please check the [example](https://github.com/opensearch-project/alerting/blob/opensearch-1.0.0.0-beta1/build-tools/merged-coverage.gradle) of a custom Gradle plugin in OpenSearch Alerting plugin, or the [sample](https://docs.gradle.org/6.6.1/samples/sample_jvm_multi_project_with_code_coverage.html) from Gradle.

#### OpenSearch Dashboards plugin
OpenSearch Dashboards mainly uses [Jest](https://jestjs.io/) for unit testing, and plugins are recommended to utilize Jest as the unit testing framework.

To generate code coverage report under Jest framework, adding the flag `--coverage` in the command to run unit tests. See [here](https://github.com/opensearch-project/security-dashboards-plugin/blob/opensearch-1.0.0.0-beta1/.github/workflows/unit-test.yml#L67) for an example in the CI workflow.

### Code Coverage Reporting

> OpenSearch plugins are using [Codecov](https://about.codecov.io/) for code coverage reporting and analysis. The dashboard can be seen [here](https://app.codecov.io/gh/opensearch-project/).

Reviewer: Can we use the same tool for reporting Dashboard plugins?

Author: Yeah, like security dahsboards plugin has already used Codecov to report the code coverage. The "OpenSearch plugins" in my words refers to the whole opensearch proejct. What do you think? 😂 To say "OpenSearch and OpenSearch Dashboards plugins" or "plugins of OpenSearch project" explicitly?

Reviewer: Oh thats nice. May say something like OpenSearch/Dashboards or OpenSearch and OpenSearch Dashboards plugins or plugins in OpenSearch project also works. 
What ever you prefer, we just have to send a message that it works for all of them like magic :)

Author: Haha, good idea. Thanks. I modified "OpenSearch plugins" to "OpenSearch/Dashboards plugins" as you suggested.
